[9.x] Move maintenance mode logic

wimulkeman opened the pull request:
Moving the maintenance mode logic to its own class. This enables users to extend/overwrite the class with their own logic.

Overwriting the logic may be needed if for example the application is run on multiple servers. In this case the developer would want to extend the storage logic for the maintenance mode to implement a solution that can change the maintenance state for all the servers at once and reverse the maintenance mode with one command.

By separating the maintenance mode logic the framework opens the door for packages that could handle the maintenance mode through different mechanisms like using the cache (as proposed in a blog https://www.raypold.com/posts/laravel-distributed-maintenance-mode/).

The need for this was shown within the discussions laravel/ideas#1944 and #36474.

This PR is a follow-up to the PR https://github.com/laravel/framework/pull/40059 with suggested changes and pointed to the master instead of the 8.x branch because it contains backward incompatible changes.

Commit:
chore (maintenance): move maintenance mode logic

Moving the maintenance mode logic to its own class. This enables users to extend/overwrite the class with their own logic.

Overwriting the logic may be needed if for example the application is running on multiple servers.

Fixes #36474

Commit:
refactor (maintenance): add isUp method

Adding the isUp method to check if the application is up or is down for maintenance.

Relates to #36474

Commit:
doc (maintenance): document code

Extend the documentation about what the code does. Also use FQN.

Relates to #36474

Commit:
fix (maintenance): correctly check if in maintenance

Relates to #36474

Commit:
refactor (maintenance): use make method

Use the make method instead of the app function to retrieve an instance of the MaintenanceMode container.

Relates to #36474

Commit:
doc (maintenance): update docblock

Update the docblock with the newly added method param.

Relates to #36474

Commit:
doc (maintenance): update docblock

Update the docblock with the newly added method param.

Relates to #36474

driesvints commented on src/Illuminate/Maintenance/MaintenanceMode.php:
```diff
@@ -0,0 +1,43 @@
+    {
+        return storage_path('framework/down');
+    }
+}
```
Please move this class into the Foundation namespace. I don't think it makes sense to create an entire new component for this. Thanks

wimulkeman replied:
Done. Thanks for reviewing the code

Commit:
style (doc): update documentation style

Updating documentation style to comply to the style used within the Laravel codebase.

Commit:
style (doc): update documentation style

Updating documentation style to comply to the style used within the Laravel codebase.

Commit:
refactor (maintenance): change namespace

Move the class from its own namespace to the Foundation namespace.

Relates to #36474

Commit:
style (doc): update documentation style

Updating documentation style to comply to the style used within the Laravel codebase.

Commit:
style (doc): update documentation style

Updating documentation style to comply to the style used within the Laravel codebase.

driesvints commented on src/Illuminate/Foundation/Http/Middleware/PreventRequestsDuringMaintenance.php:
```diff
@@ -23,15 +24,24 @@ class PreventRequestsDuringMaintenance
     protected $except = [];
 
+    /**
+     * The maintenance mode implementation. Can be used to check
+     * if the application is under maintenance.
```
```suggestion
     * The MaintenanceMode instance.
```

driesvints commented on src/Illuminate/Foundation/Http/Middleware/PreventRequestsDuringMaintenance.php:
```diff
@@ -23,15 +24,24 @@ class PreventRequestsDuringMaintenance
      *
      * @param  \Illuminate\Contracts\Foundation\Application  $app
-     * @return void
+     * @param  \Illuminate\Foundation\MaintenanceMode  $maintenanceMode
```
```suggestion
     * @param  \Illuminate\Foundation\MaintenanceMode  $maintenanceMode
     * @return void
```

driesvints replied:
I wonder if we can add a `maintenance` method to the `Application` class that returns the `Maintenance` instance? That way we don't need an extra dependency here. And you can call `$this->app->maintenance()->` for these calls.

wimulkeman replied:
That would indeed remove the dependency and make it more available within the application as a whole. I will add it to the Application class

driesvints commented on src/Illuminate/Foundation/MaintenanceMode.php:
```diff
@@ -0,0 +1,45 @@
+
+use function storage_path;
+
+class MaintenanceMode
```
Maybe this sounds better as `Maintenance`?

driesvints commented on src/Illuminate/Foundation/MaintenanceMode.php:
```diff
@@ -0,0 +1,45 @@
+        return json_decode(file_get_contents($this->getDownFilePath()), true);
+    }
+
+    private function getDownFilePath(): string
```
Add DocBlocks to all these method please.

wimulkeman committed:
Update src/Illuminate/Foundation/Http/Middleware/PreventRequestsDuringMaintenance.php

Co-authored-by: Dries Vints <dries@vints.io>

wimulkeman committed:
Update src/Illuminate/Foundation/Http/Middleware/PreventRequestsDuringMaintenance.php

Co-authored-by: Dries Vints <dries@vints.io>

Commit:
refactor (maintenance): move to application

Create an application methode to retrieve the MaintenanceMode instance. This method reduces the need to fetch the MaintenanceMode with the usage of an additional argument in the middleware.

Relates to #36474

driesvints commented on src/Illuminate/Foundation/Console/DownCommand.php:
```diff
@@ -42,21 +43,19 @@ class DownCommand extends Command
     {
         try {
-            if (is_file(storage_path('framework/down'))) {
+            if ($maintenanceMode->isDown()) {
```
```suggestion
            if ($this->laravel->maintenanceMode()->isDown()) {
```

wimulkeman replied:
@driesvints You were one step ahead of me :). I've replace the DI usage with the application method. That begins to beg the question, now the arguments have been removed from the Middleware and the commands, should this PR be retarget to the 8.x branch instead of the master branch?

driesvints commented on src/Illuminate/Foundation/Console/DownCommand.php:
```diff
@@ -42,21 +43,19 @@ class DownCommand extends Command
-                storage_path('framework/down'),
-                json_encode($this->getDownFilePayload(), JSON_PRETTY_PRINT)
-            );
+            $maintenanceMode->down($this->getDownFilePayload());
```
```suggestion
            $this->laravel->maintenanceMode()->down($this->getDownFilePayload());
```

driesvints commented on src/Illuminate/Foundation/Console/UpCommand.php:
```diff
@@ -34,18 +35,19 @@ class UpCommand extends Command
     {
         try {
-            if (! is_file(storage_path('framework/down'))) {
+            if ($maintenanceMode->isUp()) {
```
```suggestion
            if ($this->laravel->maintenanceMode()->isUp()) {
```

driesvints commented on src/Illuminate/Foundation/Console/UpCommand.php:
```diff
@@ -34,18 +35,19 @@ class UpCommand extends Command
             }
 
-            unlink(storage_path('framework/down'));
+            $maintenanceMode->up();
```
```suggestion
            $this->laravel->maintenanceMode()->up();
```

Commit:
doc (maintenance): add DockBlock

Adding documentation for the newly added methods.

Relates to #36474

Commit:
refactor (maintenance): use application method

Use the application method instead of using autowiring. The application method provides the MaintenanceMode instance which removes the requirement to inject the class in the handler method.

Relates to #36474

wimulkeman commented:
Because of the changes made the PR no longer introduces new method arguments to existing code. Should this PR shill be pointed at master (9.x) or instead to the 8.x branch?

driesvints commented on src/Illuminate/Foundation/MaintenanceMode.php:
```diff
@@ -0,0 +1,76 @@
+     */
+    public function up(): void
+    {
+        if ($this->isDown() === false) {
+            return;
+        }
+
+        unlink($this->getDownFilePath());
```
```suggestion
        if ($this->isDown() !== false) {
            unlink($this->getDownFilePath());
        }
```

driesvints commented on src/Illuminate/Foundation/MaintenanceMode.php:
```diff
@@ -0,0 +1,76 @@
+     *
+     * @return string
+     */
+    private function getDownFilePath(): string
```
```suggestion
    protected function getDownFilePath(): string
```

wimulkeman committed:
Update src/Illuminate/Foundation/MaintenanceMode.php

Co-authored-by: Dries Vints <dries@vints.io>

driesvints commented:
> Because of the changes made the PR no longer introduces new method arguments to existing code. Should this PR shill be pointed at master (9.x) or instead to the 8.x branch?

@wimulkeman you're totally correct. Feel free to re-send this to 8.x, thanks! Good job on the PR so far 👍

One thing that can very much help this PR get merged is explain the use case a bit more thoroughly in your PR description. So state clearly why this is needed (multiple servers but also different stores). 

Thanks!

wimulkeman committed:
Update src/Illuminate/Foundation/MaintenanceMode.php

Co-authored-by: Dries Vints <dries@vints.io>

seriquynh commented:
This PR no longer introduces new method arguments to existing code but it adds a new method into `Application` contract. So, it's still a breaking-change PR and should be sent to 9.x.

seriquynh commented on src/Illuminate/Foundation/MaintenanceMode.php:
```diff
@@ -0,0 +1,74 @@
+     */
+    public function up(): void
+    {
+        if ($this->isDown() !== false) {
```
`isDown` method always returns a bool value. Why don't you use just `$this->isDown()`?

```php
    public function up(): void
    {
        if ($this->isDown()) {
            unlink($this->getDownFilePath());
        }
    }
```

driesvints replied:
Ah yeah my bad, I said IsUp first but it should indeed be isDown.

wimulkeman replied:
Good point. Changed it in the code

Commit:
refactor (maintenance): remove comparison

Remove the comparison with the false value. The method always returns a bool which only allows a true or false return value. This make the comparison redundant.

Relates to #36474

driesvints reviewed (approved):
I like the PR. I'd maybe even make some proxy methods on `Application` for `isUp`, `isDown`, `up` and `down` to make the code even cleaner but let's see what @taylorotwell says.

Commit:
chore (code): remove unused import

Remove the unused import. Because the FQN is used in the DocBlock the import has become redundant.

ziadoz commented:
This looks like really useful feature. Would it be worth adding a contract/interface for the `MaintenanceMode` class itself?

Commit:
refactor (maintenance): add interface

Add an interface under the contracts to enable developers to easily extend the MaintenanceMode class.

Commit:
style (doc): update documentation style

Updating documentation style to comply to the style used within the Laravel codebase.

driesvints commented on src/Illuminate/Foundation/Application.php:
```diff
@@ -1105,14 +1105,24 @@ public function addAbsoluteCachePathPrefix($prefix)
+     */
+    public function maintenanceMode()
+    {
+        return $this->make(MaintenanceMode::class);
```
Now that you have a contract, it's probably better to resolve that. Otherwise you won't be able to overwrite the concrete implementation. We'll need to bind the contract to the container.

wimulkeman replied:
Check. I've now bind the contract to the default implementation in the src/Illuminate/Foundation/Providers/FoundationServiceProvider.php. I was not totally sure if thats the correct place to do it.

Commit:
refactor (maintenance): bind instance to contract

Bind the default MaintenanceMode class to the contract reference in the service container.

This enables users to set-up their own contract binding to overwrite the default Maintenance Mode behavior.

Commit:
style (doc): update documentation style

Updating documentation style to comply to the style used within the Laravel codebase.

axlon commented on src/Illuminate/Foundation/Providers/FoundationServiceProvider.php:
```diff
@@ -94,6 +97,16 @@ public function registerRequestSignatureValidation()
     }
 
+    /**
+     * Register de MaintenanceMode service through a contract.
```
```suggestion
     * Register the MaintenanceMode service through a contract.
```

wimulkeman committed:
Update src/Illuminate/Foundation/Providers/FoundationServiceProvider.php

Co-authored-by: Choraimy Kroonstuiver <3661474+axlon@users.noreply.github.com>

taylorotwell committed:
formatting and renaming

taylorotwell commented:
Can you allow me to push to this branch so I can update the PR? Allow edits from maintainers.

wimulkeman commented:
> taylorotwell

I've given you permission to the repository. I couldn't find the setting on this PR page, so I gave you the needed permissions instead on the forked repo. Is that sufficient?

taylorotwell commented:
Thanks for contributing to Laravel! ❤️

driesvints commented:
Thanks for your PR @wimulkeman 👍
